Feature: draw percentiles in violinplot

sasoripathos opened the pull request:
## PR Summary
Resolve #8532

Added an optional argument, a list for percentiles, to violinplot function. User now can specify the percentiles they want to draw for each violin.

I'm opening this as part of coursework from University of Toronto Scarborough's CSCD01 taught by Dr. Anya Tafliovich @atafliovich. (https://www.utsc.utoronto.ca/~atafliovich/cscd01/index.html)
## PR Checklist

- [X] Has Pytest style unit tests
- [X] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [X] New features are documented, with examples if plot related
- [X] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

dstansby commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -7926,6 +7926,11 @@ def violinplot(self, dataset, positions=None, vert=True, widths=0.5,
           If `True`, will toggle rendering of the medians.
 
+        percentiles : array-like, default=None
+          If not None, set a list of float number in interval [0, 100] for
```
```suggestion
          If not None, set a list of floats in interval [0, 100] for
```

dstansby commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8139,6 +8160,20 @@ def violin(self, vpstats, positions=None, vert=True, widths=0.5,
 
+        # Render percentile values
+        if percentiles.size > 0:
+            # Recalculate the pmins and pmaxes for percentiles
```
Could you expand this comment a little? What does `pmin` and `pmax` mean?

dstansby commented on lib/matplotlib/cbook/__init__.py:
```diff
@@ -1431,7 +1431,7 @@ def _reshape_2D(X, name):
 
 
-def violin_stats(X, method, points=100):
+def violin_stats(X, method, percentiles=None, points=100):
```
Could you add the new keyword argument to the end of the signature instead, so people calling it like `(X, method, 100)` with no keywords don't have to change their code.

dstansby commented on lib/matplotlib/cbook/__init__.py:
```diff
@@ -1451,6 +1451,11 @@ def violin_stats(X, method, points=100):
         return a vector of the values of the KDE evaluated at the values
         specified in coords.
 
+    percentiles : array-like, default = None
```
Could you specify the shape this array should be.

sasoripathos commented:
Hi @dstansby,
I have updated documentation as your suggestions and also removed the extra image-compare test cases. Please have a check!

anntzer commented:
fwiw I would prefer this to use quantiles (https://docs.scipy.org/doc/numpy/reference/generated/numpy.quantile.html) rather than percentiles (https://docs.scipy.org/doc/numpy/reference/generated/numpy.percentile.html).

QuLogic commented:
Please rebase to remove the extra test images entirely, or please squash merge this PR when it's accepted.

sasoripathos commented:
@anntzer @timhoffm I tried to use numpy.quantile, and I got error only in python3.6 (as the failing test result in commit a3912a9) which says "numpy does not have attribute quantile". After investigation, I found numpy.quantile is newly added in numpy version 1.15.0, but currently in `requirements/testing/travis36minver.txt` numpy version is specified as 1.11.0. 

So if we want to use numpy.quantile instead of numpy.percentile, I think the version number for numpy in `requirements/testing/travis36minver.txt` should be changed. Is that acceptable? Or I can revert back to the version using numpy.percentile.

timhoffm commented:
I would use quantile in the API and implement it using np.percentile(100 * q).

sasoripathos commented:
I updated the code following your suggestion - using quantile in the API and implement it using np.percentile. Is it ok to merge?

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -7880,14 +7880,14 @@ def matshow(self, Z, **kwargs):
     def violinplot(self, dataset, positions=None, vert=True, widths=0.5,
                    showmeans=False, showextrema=True, showmedians=False,
-                   points=100, bw_method=None):
+                   percentiles=None, points=100, bw_method=None):
```
Name must be changed to `quantiles`. The implementation correctly implements quantile-behavior.

sasoripathos replied:
Code and documentation updated. Please have a check!

timhoffm commented:
Not found too much on the quantile vs. percentile topic. Single source of information is [ggplot](https://ggplot2.tidyverse.org/reference/geom_violin.html), which uses quantiles. So, I think it's right for us to prefer quantiles as well.

timhoffm commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -2781,6 +2783,48 @@ def test_violinplot_bad_widths():
+
+    # Different size quantile list and plots
+    with pytest.raises(ValueError):
+        ax.violinplot(data, quantiles=[[0.1, 0.2], [0.5]])
```
Actually, this API doesn't quite make sense. You allow `quantiles=[[0.1, 0.9], [0.2, 0.8]]` but not `quantiles=[[0.1, 0.2], [0.5]]`.

Either there is just a simple case that has the same quantiles for every postion. Or one can freely define the quantiles per position. Just allowing the same number of quantiles but with different values is an odd cut.

That said, I'm unclear if there is a practical need for different quantiles per position. It prevents comparability of the quantile lines between positions.

sasoripathos replied:
I don't think it is odd. I feel it perfectly makes sense to allow user draw different quantiles for different violins. I think in this implementation user still can draw same quantiles for each violin easily. But if only support same quantiles for every postion, it would be hard if users really want to draw different quantiles for different violins.

timhoffm replied:
But if the quantiles can be different, why should you prohibit different numbers of quantiles between violins?

sasoripathos replied:
No, this example is user want to plot only one violin, but given two list of percentiles.

sasoripathos replied:
If you look at the updated code in example folder, my design accepts two violins have different number and value of quantiles. (ie. [0.1, 0.2] for first violin, [0.5, 0.7, 0.9] for second violin)

timhoffm replied:
Sorry, misinterpreted the test. Can you please use the same number of quantiles for both lists. Makes it hopefully more obvious that the number of elements in the sublists is not the issue to test.

sasoripathos replied:
Sure, I will update it with some other changes you mentioned/suggested.

sasoripathos commented on lib/matplotlib/cbook/__init__.py:
```diff
@@ -1469,6 +1475,7 @@ def violin_stats(X, method, points=100):
         - median: The median value for this column of data.
         - min: The minimum value for this column of data.
         - max: The maximum value for this column of data.
+        - quantiles: The quantile values for this column of data.
```
I am confused. If user directly use this function as API, my code only extends the results that user can get. If this extension is considered as a "break", then I don't think it is necessary to make qunatiles key optional.

sasoripathos replied:
If user never directly use this function, then it is only called by violinplot(). Should this be counted as a API though?

timhoffm replied:
Matplotlib is extensively used out in the wild. By experience, for every remote but possible way of using the official API, there's someone out there, who actually has code with it. Therefore, we've become quite sensitive to possible API breaks. They will pop up as bug reports sooner or later. So we have to at least document the changes.

Tuple unpacking of the result will break with the change:
~~~
coords, vals, mean, median, min, max = violin_stats()
~~~

sasoripathos replied:
I agree to have a document. But I feel it is odd to make quantiles optional since the current code in violin_stats doesn't care whether those value will be drawn out.

timhoffm replied:
I'm sorry, I misread the documentation here. It's a list of dictionaries with one dictionary per input dataset. Adding keys to these dicts is not breaking the API. So the above API change note is not necessary and should be removed again. Sorry for the noise.

sasoripathos commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -8106,6 +8126,7 @@ def violin(self, vpstats, positions=None, vert=True, widths=0.5,
             mins.append(stats['min'])
             maxes.append(stats['max'])
             medians.append(stats['median'])
+            quantiles = np.concatenate((quantiles, stats['quantiles']))
```
2D is not supported by the function prep_lines (typically vlines() and hlines()). In other word, these two functions requires the input at most 1D

sasoripathos commented:
@timhoffm I updated the docstring, code and test cases based on your suggestions. I also added a .rst doc in next_api_change folder. Please have a check.

sasoripathos commented:
@timhoffm API change doc removed.

timhoffm commented:
Thanks! And congratulations on your first contribution to matplotlib!
